#!/usr/bin/env python
# Copyright (c) 2020 Iqua Robotics SL - All Rights Reserved
#
# This file is subject to the terms and conditions defined in file
# 'LICENSE.txt', which is part of this source code package.


PACKAGE = "cola2_control"

from dynamic_reconfigure.parameter_generator_catkin import *

gen = ParameterGenerator()

gen.add("enable_thrusters", bool_t, 0,  "Enable thrusters", True)

gen.add("max_velocity_x",   double_t,   0, "Max velocity X produced by the pose controller", 0.3, 0.0, 0.7)
gen.add("max_velocity_y",   double_t,   0, "Max velocity Y produced by the pose controller", 0.2, 0.0, 0.3)
gen.add("max_velocity_z",   double_t,   0, "Max velocity Z produced by the pose controller", 0.2, 0.0, 0.5)
gen.add("max_velocity_roll",   double_t,   0, "Max velocity Roll produced by the pose controller", 0.0, 0.0, 0.0)
gen.add("max_velocity_pitch",   double_t,   0, "Max velocity Pitch produced by the pose controller", 0.1, 0.0, 0.3)
gen.add("max_velocity_yaw",   double_t,   0, "Max velocity Yaw produced by the pose controller", 0.5, 0.0, 1.0)


gen.add("p_surge_kp",   double_t,   0, "Pose controller Surge KP", 0.15,  0.0, 20.0)
gen.add("p_surge_ti",   double_t,   0, "Pose controller Surge TI", 0.0,  0.0, 20.0)
gen.add("p_surge_td",   double_t,   0, "Pose controller Surge TD", 0.0,  0.0, 20.0)
gen.add("p_surge_i_limit",   double_t,   0, "Pose controller Surge I limit", 0.0,  0.0, 1.0)
gen.add("p_surge_fff",  double_t,   0, "Pose controller Surge feed forward force", 0.0,  0.0, 1.0)

gen.add("p_sway_kp",   double_t,   0, "Pose controller Sway KP", 0.2,  0.0, 20.0)
gen.add("p_sway_ti",   double_t,   0, "Pose controller Sway TI", 0.0,  0.0, 20.0)
gen.add("p_sway_td",   double_t,   0, "Pose controller Sway TD", 0.0,  0.0, 20.0)
gen.add("p_sway_i_limit",   double_t,   0, "Pose controller Sway I limit", 0.0,  0.0, 1.0)
gen.add("p_sway_fff",  double_t,   0, "Pose controller Sway feed forward force", 0.0,  0.0, 1.0)

gen.add("p_heave_kp",   double_t,   0, "Pose controller Heave KP", 0.5,  0.0, 20.0)
gen.add("p_heave_ti",   double_t,   0, "Pose controller Heave TI", 0.0,  0.0, 20.0)
gen.add("p_heave_td",   double_t,   0, "Pose controller Heave TD", 0.0,  0.0, 20.0)
gen.add("p_heave_i_limit",   double_t,   0, "Pose controller Heave I limit", 0.0,  0.0, 1.0)
gen.add("p_heave_fff",  double_t,   0, "Pose controller Heave feed forward force", 0.0,  0.0, 1.0)

gen.add("p_roll_kp",   double_t,   0, "Pose controller Roll KP", 0.0,  0.0, 20.0)
gen.add("p_roll_ti",   double_t,   0, "Pose controller Roll TI", 0.0,  0.0, 20.0)
gen.add("p_roll_td",   double_t,   0, "Pose controller Roll TD", 0.0,  0.0, 20.0)
gen.add("p_roll_i_limit",   double_t,   0, "Pose controller Roll I limit", 0.0,  0.0, 1.0)
gen.add("p_roll_fff",  double_t,   0, "Pose controller Roll feed forward force", 0.0,  0.0, 1.0)

gen.add("p_pitch_kp",   double_t,   0, "Pose controller Pitch KP", 1.0,  0.0, 20.0)
gen.add("p_pitch_ti",   double_t,   0, "Pose controller Pitch TI", 0.0,  0.0, 20.0)
gen.add("p_pitch_td",   double_t,   0, "Pose controller Pitch TD", 0.0,  0.0, 20.0)
gen.add("p_pitch_i_limit",   double_t,   0, "Pose controller Pitch I limit", 0.0,  0.0, 1.0)
gen.add("p_pitch_fff",  double_t,   0, "Pose controller Pitch feed forward force", 0.0,  0.0, 1.0)

gen.add("p_yaw_kp",   double_t,   0, "Pose controller Yaw KP", 1.5,  0.0, 20.0)
gen.add("p_yaw_ti",   double_t,   0, "Pose controller Yaw TI", 0.0,  0.0, 20.0)
gen.add("p_yaw_td",   double_t,   0, "Pose controller Yaw TD", 0.0,  0.0, 20.0)
gen.add("p_yaw_i_limit",   double_t,   0, "Pose controller Yaw I limit", 0.0,  0.0, 1.0)
gen.add("p_yaw_fff",  double_t,   0, "Pose controller Yaw feed forward force", 0.0,  0.0, 1.0)



gen.add("t_surge_kp",   double_t,   0, "Twist controller Surge KP", 5.0,  0.0, 20.0)
gen.add("t_surge_ti",   double_t,   0, "Twist controller Surge TI", 0.0,  0.0, 20.0)
gen.add("t_surge_td",   double_t,   0, "Twist controller Surge TD", 0.0,  0.0, 20.0)
gen.add("t_surge_i_limit",   double_t,   0, "Twist controller Surge I limit", 0.0,  0.0, 1.0)
gen.add("t_surge_fff",  double_t,   0, "Twist controller Surge feed forward force", 0.0,  0.0, 1.0)

gen.add("t_sway_kp",   double_t,   0, "Twist controller Sway KP", 0.2,  0.0, 20.0)
gen.add("t_sway_ti",   double_t,   0, "Twist controller Sway TI", 0.0,  0.0, 20.0)
gen.add("t_sway_td",   double_t,   0, "Twist controller Sway TD", 0.0,  0.0, 20.0)
gen.add("t_sway_i_limit",   double_t,   0, "Twist controller Sway I limit", 0.0,  0.0, 1.0)
gen.add("t_sway_fff",  double_t,   0, "Twist controller Sway feed forward force", 0.0,  0.0, 1.0)

gen.add("t_heave_kp",   double_t,   0, "Twist controller Heave KP", 0.5,  0.0, 20.0)
gen.add("t_heave_ti",   double_t,   0, "Twist controller Heave TI", 0.0,  0.0, 20.0)
gen.add("t_heave_td",   double_t,   0, "Twist controller Heave TD", 0.0,  0.0, 20.0)
gen.add("t_heave_i_limit",   double_t,   0, "Twist controller Heave I limit", 0.0,  0.0, 1.0)
gen.add("t_heave_fff",  double_t,   0, "Twist controller Heave feed forward force", 0.0,  0.0, 1.0)

gen.add("t_roll_kp",   double_t,   0, "Twist controller Roll KP", 0.0,  0.0, 20.0)
gen.add("t_roll_ti",   double_t,   0, "Twist controller Roll TI", 0.0,  0.0, 20.0)
gen.add("t_roll_td",   double_t,   0, "Twist controller Roll TD", 0.0,  0.0, 20.0)
gen.add("t_roll_i_limit",   double_t,   0, "Twist controller Roll I limit", 0.0,  0.0, 1.0)
gen.add("t_roll_fff",  double_t,   0, "Twist controller Roll feed forward force", 0.0,  0.0, 1.0)

gen.add("t_pitch_kp",   double_t,   0, "Twist controller Pitch KP", 1.0,  0.0, 20.0)
gen.add("t_pitch_ti",   double_t,   0, "Twist controller Pitch TI", 0.0,  0.0, 20.0)
gen.add("t_pitch_td",   double_t,   0, "Twist controller Pitch TD", 0.0,  0.0, 20.0)
gen.add("t_pitch_i_limit",   double_t,   0, "Twist controller Pitch I limit", 0.0,  0.0, 1.0)
gen.add("t_pitch_fff",  double_t,   0, "Twist controller Pitch feed forward force", 0.0,  0.0, 1.0)

gen.add("t_yaw_kp",   double_t,   0, "Twist controller Yaw KP", 1.5,  0.0, 20.0)
gen.add("t_yaw_ti",   double_t,   0, "Twist controller Yaw TI", 0.0,  0.0, 20.0)
gen.add("t_yaw_td",   double_t,   0, "Twist controller Yaw TD", 0.0,  0.0, 20.0)
gen.add("t_yaw_i_limit",   double_t,   0, "Twist controller Yaw I limit", 0.0,  0.0, 1.0)
gen.add("t_yaw_fff",  double_t,   0, "Twist controller Yaw feed forward force", 0.0,  0.0, 1.0)


gen.add("poly_surge_A",   double_t,   0, "Twist surge poly A + Bx + Cx^2", 0.0, -20.0, 1000.0)
gen.add("poly_surge_B",   double_t,   0, "Twist surge poly A + Bx + Cx^2", 1.0, -20.0, 1000.0)
gen.add("poly_surge_C",   double_t,   0, "Twist surge poly A + Bx + Cx^2", 0.0, -20.0, 1000.0)

gen.add("poly_sway_A",   double_t,   0, "Twist sway poly A + Bx + Cx^2", 0.0, -20.0, 1000.0)
gen.add("poly_sway_B",   double_t,   0, "Twist sway poly A + Bx + Cx^2", 1.0, -20.0, 1000.0)
gen.add("poly_sway_C",   double_t,   0, "Twist sway poly A + Bx + Cx^2", 0.0, -20.0, 1000.0)

gen.add("poly_heave_A",   double_t,   0, "Twist heave poly A + Bx + Cx^2", 0.0, -20.0, 1000.0)
gen.add("poly_heave_B",   double_t,   0, "Twist heave poly A + Bx + Cx^2", 1.0, -20.0, 1000.0)
gen.add("poly_heave_C",   double_t,   0, "Twist heave poly A + Bx + Cx^2", 0.0, -20.0, 1000.0)

gen.add("poly_roll_A",   double_t,   0, "Twist roll poly A + Bx + Cx^2", 0.0, -20.0, 1000.0)
gen.add("poly_roll_B",   double_t,   0, "Twist roll poly A + Bx + Cx^2", 1.0, -20.0, 1000.0)
gen.add("poly_roll_C",   double_t,   0, "Twist roll poly A + Bx + Cx^2", 0.0, -20.0, 1000.0)

gen.add("poly_pitch_A",   double_t,   0, "Twist pitch poly A + Bx + Cx^2", 0.0, -20.0, 1000.0)
gen.add("poly_pitch_B",   double_t,   0, "Twist pitch poly A + Bx + Cx^2", 1.0, -20.0, 1000.0)
gen.add("poly_pitch_C",   double_t,   0, "Twist pitch poly A + Bx + Cx^2", 0.0, -20.0, 1000.0)

gen.add("poly_yaw_A",   double_t,   0, "Twist yaw poly A + Bx + Cx^2", 0.0, -20.0, 1000.0)
gen.add("poly_yaw_B",   double_t,   0, "Twist yaw poly A + Bx + Cx^2", 1.0, -20.0, 1000.0)
gen.add("poly_yaw_C",   double_t,   0, "Twist yaw poly A + Bx + Cx^2", 0.0, -20.0, 1000.0)


gen.add("max_wrench_X",   double_t,   0, "Max wrench X direction", 100.0, 0.0, 300.0)
gen.add("max_wrench_Y",   double_t,   0, "Max wrench Y direction", 100.0, 0.0, 300.0)
gen.add("max_wrench_Z",   double_t,   0, "Max wrench Z direction", 100.0, 0.0, 300.0)
gen.add("max_wrench_Roll",   double_t,   0, "Max wrench Roll direction", 100.0, 0.0, 300.0)
gen.add("max_wrench_Pitch",   double_t,   0, "Max wrench Pitch direction", 100.0, 0.0, 300.0)
gen.add("max_wrench_Yaw",   double_t,   0, "Max wrench Yaw direction", 100.0, 0.0, 300.0)


gen.add("max_force_thruster_positive",   double_t,   0, "Max force thruster positive", 100.0, 0.0, 400.0)
gen.add("max_force_thruster_negative",   double_t,   0, "Max force thruster negative", 100.0, 0.0, 400.0)
gen.add("thruster_distance_yaw",   double_t,   0, "Thruster distance yaw", 0.51, 0.0, 1.5)


exit(gen.generate(PACKAGE, "cola2_control", "controller"))
